lib/Containers/ResourceManager.h

```diff
@@ -0,0 +1,207 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2025 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Business Source License 1.1 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     https://github.com/arangodb/arangodb/blob/devel/LICENSE
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Max Neunhoeffer
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "Basics/cpu-relax.h"
+
+#include <atomic>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <random>
+#include <type_traits>
+#include <vector>
+
+namespace arangodb {
+
+template<typename T>
+class ResourceManager {
+ private:
+  // Alignment for cache line to prevent false sharing
+  struct alignas(64) EpochSlot {
+    std::atomic<uint64_t> epoch{0};
+  };
+
+  // Resource management
+  std::atomic<T*> current_resource;
+  std::atomic<uint64_t> global_epoch{1};  // Start at 1, 0 means "not reading"
+  std::mutex writer_mutex;
+
+  // Epoch slots for reader tracking
+  std::vector<EpochSlot> epoch_slots;
+
+  // Thread-local storage for slot index
+  static thread_local std::optional<size_t> thread_slot_index;
+
+  // Constants
+  static constexpr size_t EPOCH_SLOTS = 128;
```

Comment:
Do we want to make this a parameter? This bounds the maximal number of threads that can work at the same time on the bounded list, otherwise the additional thread will wait until there is an open slot. Could this happen often?

Reply:
This is a good question. In principle, this could happen if we use well more than 128 threads. But even then machines with more than 128 cores are very rare still. And even on such a machine it is unlikely to cause real problems any time soon, as long as readers are fast. For our application, this should be totally fine forever, therefore I would like to keep it a constant for now. If we have other applications of this class later, it is easy to change later.